test/gateway/eip20_cogateway/revert_redeem.js

```diff
@@ -0,0 +1,364 @@
+  it('should emit RevertRedeemDeclared event', async function () {
+
+    let tx = await eip20CoGateway.revertRedeem(
+      redeemParams.messageHash,
+      { from: redeemParams.redeemer, value: redeemParams.penalty },
+    );
+
+    let event = EventDecoder.getEvents(tx, eip20CoGateway);
+    let eventData = event.RevertRedeemDeclared;
+
+    assert.isDefined(
+      event.RevertRedeemDeclared,
+      'Event `RevertRedeemDeclared` must be emitted.',
+    );
+
+    assert.strictEqual(
+      eventData._messageHash,
+      redeemParams.messageHash,
+      `Message hash from the event must be equal to ${redeemParams.messageHash}.`,
+    );
+
+    assert.strictEqual(
+      eventData._redeemer,
+      redeemParams.redeemer,
+      `Redeemer address from the event must be equal to ${redeemParams.redeemer}.`,
+    );
+
+    assert.strictEqual(
+      eventData._redeemerNonce.eq(redeemParams.nonce),
+      true,
+      `Redeemer nonce ${eventData._redeemerNonce.toString(10)} from the event must be equal to ${redeemParams.nonce.toString(10)}.`,
+    );
+
+    assert.strictEqual(
+      eventData._amount.eq(redeemParams.amount),
+      true,
+      `Redeem amount ${eventData._amount.toString()} from the event must be equal to ${redeemParams.amount.toString(10)}.`,
+    );
+
+  });
+
+  it('redeemer should pay the penalty', async function () {
+
+    let eip20CoGatewayBaseBalance = new BN(
+      await web3.eth.getBalance(eip20CoGateway.address),
+    );
+
+    let redeemerBaseBalance = new BN(
+      await web3.eth.getBalance(redeemParams.redeemer),
+    );
+
+    let eip20CoGatewayBalance = await utilityToken.balanceOf(eip20CoGateway.address);
+    let redeemerBalance = await utilityToken.balanceOf(redeemParams.redeemer);
+
+    let tx = await eip20CoGateway.revertRedeem(
+      redeemParams.messageHash,
+      { from: redeemParams.redeemer, value: redeemParams.penalty },
+    );
+
+    let eip20CoGatewayBaseFinalBalance = new BN(
+      await web3.eth.getBalance(eip20CoGateway.address),
+    );
+
+    let redeemerBaseFinalBalance = new BN(
+      await web3.eth.getBalance(redeemParams.redeemer),
+    );
+
+    let eip20CoGatewayFinalBalance = await utilityToken.balanceOf(eip20CoGateway.address);
+    let redeemerFinalBalance = await utilityToken.balanceOf(redeemParams.redeemer);
+
+    assert.strictEqual(
+      eip20CoGatewayBaseFinalBalance.eq(eip20CoGatewayBaseBalance.add(redeemParams.penalty)),
+      true,
+      `CoGateway base token balance ${eip20CoGatewayBaseFinalBalance.toString(10)} must be equal to ${eip20CoGatewayBaseBalance.add(redeemParams.penalty).toString(10)}.`,
+    );
+
+    assert.strictEqual(
+      redeemerBaseFinalBalance.eq(redeemerBaseBalance.sub(redeemParams.penalty).subn(tx.receipt.gasUsed)),
+      true,
+      `Redeemer's base token balance ${redeemerBaseFinalBalance.toString(10)} must be equal to ${redeemerBaseBalance.sub(redeemParams.penalty).subn(tx.receipt.gasUsed).toString(10)}.`,
+    );
+
+    assert.strictEqual(
```

Comment:
why do we assert token balance? This assert is out of the scope of this test.

Reply:
This asserts that the base tokens are transferred and not the EIP20 tokens.